source/common/formatter/substitution_formatter.cc

```diff
@@ -1541,6 +1551,36 @@ ProtobufWkt::Value PlainStringFormatter::formatValue(const Http::RequestHeaderMa
   return str_;
 }
 
+PlainNumberFormatter::PlainNumberFormatter(const double& num) { num_.set_number_value(num); }
+
+absl::optional<std::string> PlainNumberFormatter::format(const Http::RequestHeaderMap&,
+                                                         const Http::ResponseHeaderMap&,
+                                                         const Http::ResponseTrailerMap&,
+                                                         const StreamInfo::StreamInfo&,
+                                                         absl::string_view) const {
+  std::string str = std::to_string(num_.number_value());
+  str = str.substr(0, str.find_last_not_of('0') + 1);
+  str = str.substr(0, str.find_last_not_of('.') + 1);
+  return str;
+}
+
+ProtobufWkt::Value PlainNumberFormatter::formatValue(const Http::RequestHeaderMap&,
+                                                     const Http::ResponseHeaderMap&,
+                                                     const Http::ResponseTrailerMap&,
+                                                     const StreamInfo::StreamInfo&,
+                                                     absl::string_view) const {
+  return num_;
+}
+
+std::vector<FormatterProviderPtr>
+PlainNumberFormatter::toFormatNumberValue(const double& number_format) {
```

Comment:
Let's make this a method on StructFormatter like toFormatStringValue. It also doesn't need to take a const reference. Just a `double` is fine.

Reply:
Done.